### PR TITLE
Refine API rate limit path accounting

### LINKED ISSUE #696: max_api_capacity not working properly
### Community Note

- Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
- Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for issue followers and do not help prioritize the request
- If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Terraform Version

Terraform v1.0.6
on darwin_amd64
\+ provider registry.terraform.io/okta/okta v3.14.0

### Affected Resource(s)

- all?

### Terraform Configuration Files

```hcl
provider "okta" {
  org_name         = "..."
  base_url         = "okta.com"
  api_token        = "... redacted ..."
  max_api_capacity = 80
}
```

### Debug Output

https://gist.github.com/phi1ipp/c8b01eab5f32ec6bfe96afc418aa4ed1

### Expected Behavior

Refresh state operation should be executed as there is a lot of unused capacity. 
### Actual Behavior
Terraform pauses with a message to sleep for XX seconds to preserve the limit. When I saw a message `sleeping for 60 seconds until rate limit reset ("app-id":98/1800` and then inspected all API calls above it in the log, I was not able to find where remaining `X-Rate-Limit-Remaining` would be 98 and `X-Rate-Limit-Limit` would be 1800.

Other suspicious factor is, that `/api/v1/apps` is not 1800 for sure, even though we have increased limit for this endpoint

### Steps to Reproduce

1. `terraform apply`

### PR DESCRIPTION
**Pull request by monde**
- Fix logical flaw in path accounting for API rate limits
- Increase test coverage for path accounting

### CODE REVIEW

**Comment by monde** [comment]
@phi1ipp check out this PR

**Comment by phi1ipp** [comment]
thank you @monde , looks good to me, but what do I know?! :)

**Comment by monde** [comment]
> thank you @monde , looks good to me, but what do I know?! :)

@phi1ipp your debug log and regexp example helped me see that way!

File: okta/internal/apimutex/apimutex.go [hunk 1]
```diff
@@ -24,6 +25,15 @@ const (
 	OTHER_KEY     = "other"
 )
 
+var reAppId *regexp.Regexp
+var reGroupId *regexp.Regexp
+
+func init() {
+	reAppId = regexp.MustCompile("/api/v1/apps/[^/]+$")
```
Comment (phi1ipp) [inline okta/internal/apimutex/apimutex.go]: I liked that part, definitely was a miss on my side to construct the regex for every request

File: okta/internal/apimutex/apimutex.go [hunk 2]
```diff
@@ -138,7 +148,7 @@ func (m *APIMutex) normalizeKey(method, endPoint string) string {
 		result = EVENTS_KEY
 	case endPoint == "/api/v1/groups":
 		result = GROUPS_KEY
-	case strings.HasPrefix(endPoint, "/api/v1/groups/"):
+	case reGroupId.MatchString(endPoint):
```
Comment (phi1ipp) [inline okta/internal/apimutex/apimutex.go:145]: that's a good catch also, I totally skipped this endpoint

File: okta/internal/apimutex/apimutex.go (hunk 1 above)
Comment (bogdanprodan-okta) [reply okta/internal/apimutex/apimutex.go]: Can be simplified to just
`var reAppId = regexp.MustCompile("/api/v1/apps/[^/]+$")`

Comment (phi1ipp) [reply okta/internal/apimutex/apimutex.go]: very true!

Comment (monde) [reply okta/internal/apimutex/apimutex.go]: cleaned it up

**Review by bogdanprodan-okta — APPROVED** [review]

**Pull request merged**